### PR TITLE
test: add more tests for CRS methods

### LINKED ISSUE #261: Add more CRS tests

### PR DESCRIPTION
**Pull request by anthony-swirldslabs**
**Description**:
Adding more comprehensive tests for CRS method in the `HintsLibraryBridge`.
Also, partially addressing https://github.com/hashgraph/hedera-cryptography/issues/240 by improving the deserialization method in the `jni_util.rs`, and using the new capability in the tests.

**Related issue(s)**:

Fixes #261 

**Notes for reviewer**:
All tests should pass.

**Checklist**

- [x] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

File: cryptography/hedera-cryptography-hinTS/src/test/java/com/hedera/cryptography/hints/HintsLibraryBridgeCRSTest.java [hunk 1]
```diff
@@ -52,6 +69,24 @@ void testUpdateCRS() {
         updateAndVerifyCRS(prevCRS);
     }
 
+    @Test
+    void testUpdateCRSConstraints() {
+        assertNull(INSTANCE.updateCRS(null, CRSConstants.RANDOM));
+        assertNull(INSTANCE.updateCRS(initAndVerifyCRS(), null));
+
+        // Only byte[32] is valid for entropy input:
```
Comment (rsinha) [inline cryptography/hedera-cryptography-hinTS/src/test/java/com/hedera/cryptography/hints/HintsLibraryBridgeCRSTest.java:77]: I didn't follow why byte[32] specifically. More generally, what are the following assertions capturing?

File: cryptography/hedera-cryptography-hinTS/src/test/java/com/hedera/cryptography/hints/HintsLibraryBridgeCRSTest.java [hunk 2]
```diff
@@ -62,5 +97,44 @@ void testVerifyCRS() {
                 Arrays.copyOfRange(nextCRSandContributionProof, CRS_4_LENGTH, nextCRSandContributionProof.length);
 
         assertTrue(INSTANCE.verifyCRS(prevCRS, nextCRS, contributionProof));
+
```
Comment (rsinha) [inline cryptography/hedera-cryptography-hinTS/src/test/java/com/hedera/cryptography/hints/HintsLibraryBridgeCRSTest.java:100]: You could also consider damaging by swapping some element in the vector with another -- that said, it might be difficult to do once you have serialized the CRS.

Comment (rsinha) [reply cryptography/hedera-cryptography-hinTS/src/test/java/com/hedera/cryptography/hints/HintsLibraryBridgeCRSTest.java:100]: By modifying some bytes in the CRS, we are most likely only triggering deserialization failures. If that is the intention of these tests, then it's perfectly fine.

File: cryptography/hedera-cryptography-hinTS/src/main/rust/jni_crs.rs [hunk 3]
```diff
@@ -29,7 +29,7 @@ use crate::jni_util;
 /// # Returns
 /// *   a byte array with a serialized CRS object, or null on error
 #[no_mangle]
-pub extern "system" fn Java_com_hedera_cryptography_hints_HintsLibraryBridge_initCRS(
+pub extern "system" fn Java_com_hedera_cryptography_hints_HintsLibraryBridge__1initCRS(
```
Comment (mxtartaglia-sl) [inline cryptography/hedera-cryptography-hinTS/src/main/rust/jni_crs.rs]: Is the name change intentional? Seems that there is an extra 1 in the name

Comment (mxtartaglia-sl) [reply cryptography/hedera-cryptography-hinTS/src/main/rust/jni_crs.rs]: Never mind, I see its JNI convention for anything starting with _

File: cryptography/hedera-cryptography-hinTS/src/main/java/com/hedera/cryptography/hints/HintsLibraryBridge.java [hunk 4]
```diff
@@ -39,11 +39,26 @@ public static HintsLibraryBridge getInstance() {
...
-    public native byte[] initCRS(long signersNum);
+    public byte[] initCRS(long signersNum) {
+        // Support a degenerate case of 0 signers, or a normal case with more signers. Otherwise, error out.
+        if (signersNum < 1) {
+            return null;
+        }
+        return _initCRS(signersNum);
+    }
+
+    private native byte[] _initCRS(long signersNum);
```
Comment (mxtartaglia-sl) [inline cryptography/hedera-cryptography-hinTS/src/main/java/com/hedera/cryptography/hints/HintsLibraryBridge.java]: I understand the reasoning for having the name starting with _, but it goes against standard Java's code style. I don't think we should blindly follow conventions, but I do think we have to be consistent in all codebase, and this would be one of the few locations we name private methods with _ (to the best of my knowledge).
My proposal: public InitiateCRS vs private initCRS

Comment (anthony-swirldslabs) [reply cryptography/hedera-cryptography-hinTS/src/main/java/com/hedera/cryptography/hints/HintsLibraryBridge.java]: Fair. I'm renaming `_initCRS` to `initCRSImpl`.

File: cryptography/hedera-cryptography-hinTS/src/test/java/com/hedera/cryptography/hints/HintsLibraryBridgeCRSTest.java (hunk 1 above)
Comment (anthony-swirldslabs) [reply cryptography/hedera-cryptography-hinTS/src/test/java/com/hedera/cryptography/hints/HintsLibraryBridgeCRSTest.java:77]: I recall we discussed the random input/entropy in the past and agreed that we want to accept 256 bits of entropy for random input, which is equivalent to 32 bytes. This has been codified at: https://github.com/hashgraph/hedera-cryptography/blob/67fb03c6602864927f03efc8ca32d1172da89c4e/cryptography/hedera-cryptography-hinTS/src/main/rust/hints.rs#L41-L41

Therefore, the deserialization for the random input in native code uses a [u8; RANDOM_SIZE] array and requires the input size to match that.

Hence all these assertions.

File: cryptography/hedera-cryptography-hinTS/src/test/java/com/hedera/cryptography/hints/HintsLibraryBridgeCRSTest.java (hunk 2 above)
Comment (anthony-swirldslabs) [reply cryptography/hedera-cryptography-hinTS/src/test/java/com/hedera/cryptography/hints/HintsLibraryBridgeCRSTest.java:100]: The CRS is defined at https://github.com/hashgraph/hedera-cryptography/blob/67fb03c6602864927f03efc8ca32d1172da89c4e/cryptography/hedera-cryptography-hinTS/src/main/rust/kzg.rs#L35-L41

as a struct containing two vectors. While I don't know exactly where these vectors are in the serialized form, there's a possibility to break the serialization per se (e.g. if the damaged bytes represent the length of the vector), but there's also a possibility to damage the data elements only, which must be (I presume) numbers. I'd love to be able to damage only the numbers, so that the deserialization succeeds and we check if the actual math in our implementation detects the damage or not. But w/o knowing the exact format of the serialized data, it's a bit difficult to do that deterministically. So my best bet was to modify some random elements.

If you have a better idea, or can point to specific indices to modify, I'd be more than happy to do that. Please let me know.

Comment (rsinha) [reply cryptography/hedera-cryptography-hinTS/src/test/java/com/hedera/cryptography/hints/HintsLibraryBridgeCRSTest.java:100]: On second thought, it is better to have those tests be done on rust side.

**Comment by rsinha** [comment]
lgtm

File: cryptography/hedera-cryptography-hinTS/src/test/java/com/hedera/cryptography/hints/HintsLibraryBridgeCRSTest.java (hunk 2 above)
Comment (anthony-swirldslabs) [reply cryptography/hedera-cryptography-hinTS/src/test/java/com/hedera/cryptography/hints/HintsLibraryBridgeCRSTest.java:100]: I disagree. There's multiple benefits of having them in Java. We're building an official Java interface for this library (in the form of the Bridge classes), and we need to ensure the interfaces are well tested, including any edge cases. By having the tests written in Java, we test the entire stack of the library - from the input interface, through the JNI bindings, down to the actual native code that implements the functionality. This way we don't need to test the same corner case multiple times at multiple levels - that wouldn't be extremely bad, of course, but would still be wasteful from LOC/effort perspectives.

Additionally, since 99% of the consensus node is implemented in Java, it's easier to implement these tests in Java as we have a lot more expertise in this language than any other. So it will be easier to maintain these tests in the future. At some point, we may even decide to port (parts of) the native code to Java, in which case we'd have all the tests already in place, and they could play a role of regression tests in this case to ensure the ported code is 100% compatible with the original native one.

**Review by mxtartaglia-sl — APPROVED** [review]

**Review by andrewb1269 — APPROVED** [review]
